include/zephyr/drivers/sensor.h

```diff
@@ -394,12 +398,184 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
+ * @brief Used for iterating over data channels in the same frame via :c:struct:`sensor_decoder_api`
+ */
+typedef uint32_t sensor_channel_iterator_t;
+
+/**
+ * @brief Decodes a single raw data buffer
+ *
+ * Data buffers are provided on the :c:struct:`rtio` context that's supplied to
+ * c:func:`sensor_read`.
+ */
+struct sensor_decoder_api {
+	/**
+	 * @brief Get the number of frames in the current buffer.
+	 *
+	 * @param[in]  buffer The buffer provided on the :c:struct:`rtio` context.
+	 * @param[out] frame_count The number of frames on the buffer (at least 1)
+	 * @return 0 on success
+	 * @return <0 on error
+	 */
+	int (*get_frame_count)(const uint8_t *buffer, uint16_t *frame_count);
+
+	/**
+	 * @brief Get the timestamp associated with the first frame.
+	 *
+	 * @param[in]  buffer The buffer provided on the :c:struct:`rtio` context.
+	 * @param[out] timestamp_ns The closest timestamp for when the first frame was generated
+	 *             as attained by :c:func:`k_uptime_ticks`.
+	 * @return 0 on success
+	 * @return <0 on error
+	 */
+	int (*get_timestamp)(const uint8_t *buffer, uint64_t *timestamp_ns);
+
+	/**
+	 * @brief Get the shift count of a particular channel (multiplier)
+	 *
+	 * This value can be used by shifting the q31_t value resulting in the SI unit of the
+	 * reading. It is guaranteed that the shift for a channel will not change between frames.
+	 *
+	 * @param[in]  buffer The buffer provided on the :c:struct:`rtio` context.
+	 * @param[in]  channel_type The c:enum:`sensor_channel` to query
+	 * @param[out] shift The bit shift of the channel for this data buffer.
+	 * @return 0 on success
+	 * @return -EINVAL if the @p channel_type doesn't exist in the buffer
+	 * @return <0 on error
+	 */
+	int (*get_shift)(const uint8_t *buffer, enum sensor_channel channel_type, int8_t *shift);
+
+	/**
+	 * @brief Decode up to N samples from the buffer
+	 *
+	 * This function will never wrap frames. If 1 channel is available in the current frame and
+	 * @p max_count is 2, only 1 channel will be decoded and the frame iterator will be modified
+	 * so that the next call to decode will begin at the next frame.
+	 *
+	 * @param[in]     buffer The buffer provided on the :c:struct:`rtio` context
+	 * @param[in,out] fit The current frame iterator
+	 * @param[in,out] cit The current channel iterator
+	 * @param[out]    channels The channels that were decoded
+	 * @param[out]    values The scaled data that was decoded
+	 * @param[in]     max_count The maximum number of channels to decode.
+	 * @return
+	 */
+	int (*decode)(const uint8_t *buffer, sensor_frame_iterator_t *fit,
+		      sensor_channel_iterator_t *cit, enum sensor_channel *channels, q31_t *values,
+		      uint8_t max_count);
+};
+
+/**
+ * @typedef sensor_get_decoder_t
+ * @brief Get the decoder associate with the given device
+ *
+ * @see sensor_get_decoder for more details
+ */
+typedef int (*sensor_get_decoder_t)(const struct device *dev,
```

Comment:
I think this is ok if some rules are applied. The API struct needs to be global (not on a stack somewhere), and needs to be viewed as immutable from the user space/callers perspective.

If those hold true this should be fine. Alternatively you could copy it to caller provided memory. The benefits of the copy might be there isn't an unexpected mutation from the driver causing some hassles here.

``` c
struct sensor_decoder_api decoder;

sensor_get_decoder(sensor_dev, &decoder); 
```

doesn't seem too bad.

Reply:
Right, the goal (not yet documented) is to have these just 1 per sensor and they should always be the exact same from boot (drivers should never mutate them _EVER_).

Reply:
In any case after discussing with @ceolin this does need to copy out

Reply:
Done